경북대 BE_심규민 2주차 과제 (1단계)

sim-mer opened the pull request:
- 예외 발생 시 응답 내용에 Bean Validator에서 생성된 메시지나 자세한 예외 발생 시 발생하는 메시지도 추가하도록 변경하였습니다.
- 상품 등록과 수정 DTO를 통합하고 Bean Validation 적용 후  rename 하였고, 삭제 요청 시 사용되는 DTO도 같은 방식으로 수정했습니다.

sim-mer committed:
merge

sim-mer committed:
refactor(): 사용하지 않는 ProductRepository 구현체 제거

sim-mer committed:
refactor(controller): RestController, Controller 패키지 분리

sim-mer committed:
chore(build.gradle): Spring boot starter validation 의존성 추가

sim-mer committed:
refactor(exception): ErrorResponseDto 이동
- GlobalExceptionHandler.java 에서만 사용하므로 파일 내부로 이동

sim-mer committed:
test(): 상품 등록 기대 응답코드 201로 변경

sim-mer committed:
style(dto): dto 코드 컨벤션 변경

sim-mer committed:
refactor(dto): CreateProductDto, UpdateProductDto 통합
- RequestProductDto로 통합

sim-mer committed:
feat(exception): 에러 응답에 세부 사항 추가

sim-mer committed:
refactor(): 관리자 페이지 상품 수정 요청 AJAX로 변경, Controller의 수정 요청 제거

sim-mer committed:
refactor(): ProductDto -> ResponseProductDto로 rename

sim-mer committed:
refactor(): ProductDeleteDto 변경
- RequestProductIdsDto로 rename
- Valid 어노테이션 추가

sim-mer committed:
refactor(controller)
- Request Dto를 인자로 받는 메서드에 @Valid 추가
- Query Param -> PathVariable로 변경

sim-mer committed:
refactor() : 관리자 페이지 삭제 요청 Mapping url 변경

sim-mer committed:
test() : 테스트 코드 요청 정보 PathVariable 사용으로 변경

sim-mer committed:
fix() : conflict 해결

leaf-upper reviewed (changes requested):
안녕하세요 규민님~

**아래 기능 요구사항 구현되어있는지 확인 부탁드립니다.**

상품을 추가하거나 수정하는 경우, 클라이언트로부터 잘못된 값이 전달될 수 있다. 잘못된 값이 전달되면 클라이언트가 어떤 부분이 왜 잘못되었는지 인지할 수 있도록 응답을 제공한다.

1. 상품 이름은 공백을 포함하여 최대 15자까지 입력할 수 있다.

특수 문자
가능: ( ), [ ], +, -, &, /, _
그 외 특수 문자 사용 불가

2. "카카오"가 포함된 문구는 담당 MD와 협의한 경우에만 사용할 수 있다.

sim-mer committed:
refactor() : 상품 추가 ajax 요청으로 변경

sim-mer committed:
style() : import 정리

sim-mer committed:
feat() : 상품 이름에 문자 필터링 추가

sim-mer committed:
feat() : 상품 이름에 '카카오' 필터링 추가

sim-mer committed:
style() : Google Java Style Guide 적용

sim-mer committed:
test() : createProduct 특수문자, 카카오 실패케이스 추가

sim-mer commented:
제가 요구 사항을 제대로 못봤네요.. 수정 완료 했습니다!
### 궁금한 점
- 계층형 아키텍처에서 DTO의 사용범위는 어디까지 일까요? 현재 서비스 레이어에서 DTO to Domain, Domain to DTO를 수행하고 DTO를 반환하고 있는데 Controller에서 모든 변환이 수행된다고 했을 경우 변환 로직이 복잡할 때 여러 서비스를 의존할 수 있는 문제가 생길 수 있고, 클라이언트로 반환하지 않아도 되는 Domain의 정보가 Controller까지 넘어오는 문제도 생길 수 있습니다. 하지만 여러 종류의 컨트롤러가 한 서비스를 사용할 경우에는 DTO로 서비스를 호출하는 건 적절해 보이지 않는 것 같습니다.

sim-mer committed:
refactor() : 패키지 구조 변경
- gift 패키지 아래 business, presentation, persistence 패키지로 계층 분리
- 각 패키지 아래 dto, model과 로직 패키지 분리

leaf-upper commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -1,31 +1,61 @@
-        return createErrorResponseEntity(ErrorCode.INVALID_REQUEST);
+    public ResponseEntity<ErrorResponseDto> handleMethodArgumentTypeMismatchException(
+        MethodArgumentTypeMismatchException e) {
+        return createErrorResponseEntity(ErrorCode.INVALID_REQUEST, null);
```
null보다는 emptyMap을 사용하여, null을 최대한 사용하지 않는방식으로 코드를 작성해보는건 어떨까요?

sim-mer replied:
넵 수정했습니다!

leaf-upper commented on src/main/java/gift/business/service/ProductService.java:
```diff
@@ -1,36 +1,35 @@
 
 
-    public ProductDto getProduct(Long id) {
+    public ResponseProductDto getProduct(Long id) {
```
Service 계층에서는 Controller 계층에서 정의된, DTO를 사용하기 보다는
Service 계층 자체에서 사용하기 위한 별도 DTO를 만들어서 작업해보시건 어떠실까요?

예시를 들자면

Controller -> ResponseProductDto
Service -> ProductDto

- Service에서는 ProductDto를 반환하고, 이후 Controller에서 
ProductDto를 ResponseProductDto로 매핑하여 클라이언트에게 응답.

Controller -> Service -> Respository계층으로 DTO를 사용하거나 의존할 수 있으며, 반대방향으로는 불가
ex. 
- Controller에서는 Service계층의 Dto 사용가능
- Service에서는 Controller계층의 Dto 사용불가

sim-mer replied:
수정했습니다!

leaf-upper commented on src/main/java/gift/business/service/ProductService.java:
```diff
@@ -1,36 +1,35 @@
 
-    public Long createProduct(CreateProductDto giftDto) {
-        Product product = giftDto.toProduct();
+    public Long createProduct(RequestProductDto requestProductDto) {
```
create, update를 하는코드도 마찬가지로 Service계층에서 DTO를 정의해서 사용해보면 좋을것 같아요~

ex. 
ProductRegisterDto
ProductUpdateDto

sim-mer replied:
수정했습니다!

leaf-upper commented on src/main/java/gift/presentation/restcontroller/ProductController.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<Long> createProduct(@RequestBody @Valid RequestProductDto giftDto) {
```
위에서 말씀드린것처럼 계층별로 DTO가 존재하게 되면,
여기서 RequestProductDto를 ProductRegisterDto로 바꾸는 작업이 필요할것 같아요~

sim-mer replied:
수정했습니다!

leaf-upper commented on src/main/java/gift/presentation/dto/RequestProductDto.java:
```diff
@@ -0,0 +1,27 @@
+    String name,
+    @NotNull @Min(0) Integer price,
+    @Size(max = 255) String description,
+    @URL String imageUrl
```
URL Validation은 hibernate 패키지의 validation이 사용되었네요~ 
hibernate는 JPA의 구현체입니다. Hibernate는 DB에 save이전, 
해당 애너테이션을보고 Bean validation을 해주는 기능을 가지고 있어요.
이 기능은 스프링 Validation과 무관할수도 있습니다.

그래서 그런지 제대로 validation이 동작할지 의문이드네요
TC도 한번 작성해주시면 좋을것 같습니다~

leaf-upper replied:
요거는 제가 궁금해서 좀더 찾아봤는데 사용이 가능한것 같네요 ㅎㅎ..
ORM에서만 사용하는게 아니라 Bean Validation 표준이군요
제약사항 관련 TC만 좀더 추가해주시면 될것 같아요!

sim-mer replied:
Hibernate Validator는 Bean Validation 표준 (JSR 380, Bean Validation 2.0)의 구현체로 Hibernate ORM과는 별개인 걸로 알고 있습니다! TC는 추가했고 URL validate 잘 동작하는 것 확인했습니다!

leaf-upper replied:
👍 👍 크 완벽합니다. 고생하셨어요~!

leaf-upper reviewed (changes requested):
안녕하세요 규민님
1단계 코드 너무 잘작성해주셨고 고생많으셨습니다~ 👍 👍 

- 계층형 아키텍처에서 DTO의 사용범위는 어디까지 일까요?

이 질문에 대해서는 제가 개인적으로 생각하는바를 작성해둘게요~
Controller -> Service -> Repository 각 계층별로  Dto를 저는 두는걸 선호합니다.

Controller는 Service 혹은 Repository의 dto를 사용할순 있지만, 그 반대는 불가하도록 코드를 작성하는 편이에요.

Controller의 Dto는 작성하신코드에서 RequestProductDto가 될수 있을것 같고, 이 값은 만들어져서 Front계층에서 전달될것 같아요.
그와 마찬가지로 서비스 계층에서도 똑같이 예시로 ProductRegisterDto 같은 객체를 만들고 Controller에서 해당 객체를 만들어서, 서비스를 호출할 수 있습니다.

Repository는 별도로 DTO를 만든다기보다는 영속적으로 저장하는 Entity 그자체를 이용하시는게 좋을것 같아요. 여기선 Product가 되겠네요~

혹시 설명이 부족하다면 코멘트 부탁드리고 리뷰는 한번 확인 부탁드려요~

sim-mer committed:
fix(exception) : null 대신 emptyMap으로 변경

sim-mer commented:
@leaf-upper  핵심 비스니스 로직을 다루는 계층이 Service 레이어이기 때문에 service 계층에서는 dto보다는 도메인 모델이라는 명칭이 더 어울리고 반면에 persistence 계층에는 Entity가 일종의 persistence DTO이기에 dto라는 명칭이 더 낫다고 생각하는데 어떻게 생각하시는지 궁금합니다!

sim-mer committed:
test() : url validation tc 작성

sim-mer committed:
refactor() : product 클래스 패키지 변경
- persistence 아래 entity 패키지로 이동

sim-mer committed:
refactor() : service 계층 메서드들의 리턴을 위한 DTO인 ProductDto 생성

sim-mer committed:
refactor() : service 계층 메서드들의 매개변수를 위한 ProductRegisterDto 작성

leaf-upper reviewed (approved):
NEXTSTEP으로 리뷰요청이 오지 않았지만, 일단 1단계를 완료하신것 같아서 리뷰를 드릴게요~!
1단계 고생 많으셨고 정말 코드 잘작성해주신것 같아요! Approve 드리겠습니다.
근데 지금 조직 권한이 사라져서, Merge가 불가능한데요 ㅠ 이후에 머지는 해놓도록 하겠습니다.
다음단계 진행 부탁드려요!

--- 
아래처럼 질문 주셨는데요,

핵심 비스니스 로직을 다루는 계층이 Service 레이어이기 때문에 service 계층에서는 dto보다는 도메인 모델이라는 명칭이 더 어울리고 반면에 persistence 계층에는 Entity가 일종의 persistence DTO이기에 dto라는 명칭이 더 낫다고 생각하는데 어떻게 생각하시는지 궁금합니다!

사실 제가 잘 이해했는지는 모르겠어요~ dto는 계층간 데이터를 이동하는 객체이고, 도메인 모델이란 해당 도메인에 속해서 비즈니스로직을 수행하는 객체라고 생각해요.
ProductDto는 Dto이고, Product는 도메인모델 일것 같네요
혹시 더 궁금하신 사항 있으시면 코멘트주세요~!
